### PR TITLE
Color Manager & Themes

### PR DESCRIPTION
**Pull request by LucaHaverty**
### Description
Refactors the current color manager to have better names and support for multiple themes. A new theme system will allow users to switch between color themes and create their own or modify existing ones through the appdata folder. The themes will affect UI as well as the skybox and floor color. There will be a theme editor modal in settings to create, delete, and modify themes.

### Objectives
- [x] Create a new color manager system
- [x] Add support for themes
- [x] Set up an event system for when the theme is changed
- [x] Convert all existing color code to the improved color manager
- [x] Add theme dropdown the settings modal
- [x] Make open UI colors update when the theme is changed
- [x] UI to edit themes in runtime
- [x] UI to Create / Delete themes in runtime
- [x] Rename colors so they make more sense and add things such as floor and skybox to the color manager

### Testing Done
- Creating, deleting, and modifying of colors in a theme json file in runtime
- Deleted the themes folder in synthesis appdata and it was successfully replaced
- Switching between and editing multiple themes at the same time

[JIRA Issue (Color Manager)](https://jira.autodesk.com/browse/AARD-1556)
[JIRA Issue (Themes)](https://jira.autodesk.com/browse/AARD-1557)

### CODE REVIEW

**Review by BrandonPacewic — CHANGES_REQUESTED** [review]
I had some problems when trying to load and create a new theme.

Unity Error:

```
IndexOutOfRangeException: Index was outside the bounds of the array.
UI.Dynamic.Modals.Configuring.ThemeEditor.EditThemeModal.SetThemePref () (at Assets/Scripts/UI/Dynamic/Modals/Configuring/ThemeEditor/EditThemeModal.cs:246)
UI.Dynamic.Modals.Configuring.ThemeEditor.EditThemeModal.SelectTheme (System.Int32 index) (at Assets/Scripts/UI/Dynamic/Modals/Configuring/ThemeEditor/EditThemeModal.cs:222)
UI.Dynamic.Modals.Configuring.ThemeEditor.EditThemeModal.Create () (at Assets/Scripts/UI/Dynamic/Modals/Configuring/ThemeEditor/EditThemeModal.cs:63)
Synthesis.UI.Dynamic.DynamicUIManager.CreateModal[T] (System.Object[] args) (at Assets/Scripts/UI/Dynamic/DynamicUIManager.cs:76)
```

And my `Select a Theme` menu looked like this:

Screenshot 2023-07-11 at 9 04 08 AM

It looks like it's having problems loading the preset themes.

**Comment by LucaHaverty** [comment]
> I had some problems when trying to load and create a new theme.
> 
> Unity Error:
> 
> ```
> IndexOutOfRangeException: Index was outside the bounds of the array.
> UI.Dynamic.Modals.Configuring.ThemeEditor.EditThemeModal.SetThemePref () (at Assets/Scripts/UI/Dynamic/Modals/Configuring/ThemeEditor/EditThemeModal.cs:246)
> UI.Dynamic.Modals.Configuring.ThemeEditor.EditThemeModal.SelectTheme (System.Int32 index) (at Assets/Scripts/UI/Dynamic/Modals/Configuring/ThemeEditor/EditThemeModal.cs:222)
> UI.Dynamic.Modals.Configuring.ThemeEditor.EditThemeModal.Create () (at Assets/Scripts/UI/Dynamic/Modals/Configuring/ThemeEditor/EditThemeModal.cs:63)
> Synthesis.UI.Dynamic.DynamicUIManager.CreateModal[T] (System.Object[] args) (at Assets/Scripts/UI/Dynamic/DynamicUIManager.cs:76)
> ```
> 
> And my `Select a Theme` menu looked like this:
> 
> Screenshot 2023-07-11 at 9 04 08 AM
> It looks like it's having problems loading the preset themes.

If anyone else is still having issues with a blank theme, go into your synthesis appdata folder and delete any themes without a name or named Default.

**Review by BrandonPacewic — COMMENTED** [review]
Could we have a button that will reset our custom saved themes?

**Comment by LucaHaverty** [comment]
> Could we have a button that will reset our custom saved themes?

Do you mean reset to the defaults or reset just the most recent changes?

**Comment by BrandonPacewic** [comment]
> > Could we have a button that will reset our custom saved themes?
> 
> Do you mean reset to the defaults or reset just the most recent changes?

I mean reset all custom saved themes. So basically just remove all the files in the themes folder on your computer.

**Review by Markytt — CHANGES_REQUESTED** [review]
Could you add a cancel button to the button left corner of the theme editor panel, right now the only way to close it is by clicking the save button. 
Screenshot 2023-07-11 140411

**Review by BrandonPacewic — COMMENTED** [review]
It looks like some / all of the analytics updates are in here. Was that on purpose?

**Review by KyroVibe — CHANGES_REQUESTED** [review]
Couple things I've noticed:

1. The three dashes at the top of the MainHUD doesn't update to a new color.
2. The version text doesn't update either.
3. When the grid lines are set to black, the grid is actually grey.
4. If the themes folder doesn't exist, getting existing themes breaks.

Also I know it's kinda late to ask this, but could we switch the color selection to HSV instead of RGB?

**Comment by LucaHaverty** [comment]
@KyroVibe The three dashes on the main HUD update for me. They are set to the icon color.

**Comment by LucaHaverty** [comment]
> It looks like some / all of the analytics updates are in here. Was that on purpose?

Yeah I accidentally made this branch off of analytics instead of dev. Hunter said it's fine but we have to merge analytics before this.

**Review by BrandonPacewic — COMMENTED** [review]
Everytime I click on a panel it stays highlighted.

Screenshot 2023-07-13 at 11 00 51 AM

**Review by BrandonPacewic — COMMENTED** [review]
~~When I click on preview everything gets highlighted.~~

Screenshot 2023-07-13 at 1 58 10 PM

**Comment by Markytt** [comment]
> I like the UI and it seems to work pretty well! I did find a couple issues though
> 
> * If you try to edit a theme from the main menu after exiting from the main scene, it will throw an exception trying to create a theme, cancel the editor, or accept, because TopBar has been destroyed. The Create Theme modal will also not close the theme editor one.
> * If you create a theme in the main menu and then try to enter the main scene, it will throw a KeyNotFoundException because the "escape_menu" key isn't in controls (although I'm not sure that's specifically an issue with this PR)
> 
> Along with this, maybe an "Apply" button would be nice so you can preview changes without having to save them, and so that you can adjust the settings easily to find one you like instead of having to exit and re-enter every time you want to update the colors.

Just going to the main menu and re-entering practice mode will also cause the KeyNotFoundException 

Screenshot 2023-07-13 142015
Screenshot 2023-07-13 142030
Screenshot 2023-07-13 142053

**Comment by LucaHaverty** [comment]
Yeah it happens whenever you enter practice mode

**Comment by LucaHaverty** [comment]
@BrandonPacewic I am getting a format validation error even though I ran the formatter. It says the error is on line 35 of ColorManager.cs.

**Comment by BrandonPacewic** [comment]
@LucaHaverty Fixed it, is this pr ready for review?

**Comment by LucaHaverty** [comment]
> @LucaHaverty Fixed it, is this pr ready for review?

@BrandonPacewic Yes

File: engine/Assets/Scripts/Utilities/ColorManager/ColorManager.cs
Comment (BrandonPacewic) [inline engine/Assets/Scripts/Utilities/ColorManager/ColorManager.cs:25]: To make this more readable I think we should save the separation character to a variable.

```cs
char altSep = Path.AltDirectorySeparatorChar;
```

Other than that looks good.

**Review by KyroVibe — APPROVED** [review]
Added a quick shader fix previously not seen due to a fixed grid line color. Nice work

**Review by BrandonPacewic — APPROVED** [review]
Nice work, very impressed that you got this done so quickly!

**Review by Markytt — COMMENTED** [review]
Changing/editing the theme editor in the main menu still causes a MissingReferenceException. Other than that though, all of the other bugs appear to have been fixed. 
Screenshot 2023-07-14 093321

**Review by PepperLola — CHANGES_REQUESTED** [review]
Sorry Luca, I know you want to be done with this, but do you think you could make it so that when you delete a theme, it resets all of the values back to default? I created a theme and changed a color, then deleted the theme and the colors stayed visible even though the default theme was selected. I created a new theme and these colors were automatically applied and the default one was reset back to normal. I also think that when you create a new theme, it should automatically have the default theme colors regardless of which theme you have selected. It seems like this is an issue with the preview stuff but I'm not sure.

**Comment by LucaHaverty** [comment]
> Sorry Luca, I know you want to be done with this, but do you think you could make it so that when you delete a theme, it resets all of the values back to default? I created a theme and changed a color, then deleted the theme and the colors stayed visible even though the default theme was selected. I created a new theme and these colors were automatically applied and the default one was reset back to normal. I also think that when you create a new theme, it should automatically have the default theme colors regardless of which theme you have selected. It seems like this is an issue with the preview stuff but I'm not sure.

@PepperLola I can't seem to recreate that issue. Did you use "delete selected" or "delete all"? And did you save the theme before deleting it? And in the main menu or in practice mode?

**Comment by PepperLola** [comment]
> > Sorry Luca, I know you want to be done with this, but do you think you could make it so that when you delete a theme, it resets all of the values back to default? I created a theme and changed a color, then deleted the theme and the colors stayed visible even though the default theme was selected. I created a new theme and these colors were automatically applied and the default one was reset back to normal. I also think that when you create a new theme, it should automatically have the default theme colors regardless of which theme you have selected. It seems like this is an issue with the preview stuff but I'm not sure.
> 
> @PepperLola I can't seem to recreate that issue. Did you use "delete selected" or "delete all"? And did you save the theme before deleting it? And in the main menu or in practice mode?

This was on the main menu if that affects anything. What I did was:
1. Create new theme
2. Change primary color
3. Click preview
4. Delete selected
5. Create new theme
6. Preview/save theme
I didn't test it very extensively so it's possible that it's only the preview colors that are being carried over, but regardless I think they should automatically reset to the defaults when you create a new theme.

**Review by PepperLola — APPROVED** [review]
Looks good! The colors aren't carried over when deleting or creating new themes and I'm able to exit to the main menu and still change themes. ~~For some reason when I pulled, the AddRobotModal wasn't imported in ServerTestMode. I'm still going to approve the PR, but I would check if that's an issue on your end before merging.~~ Nevermind, you just fixed it.

**Review by Markytt — APPROVED** [review]
Looks like there are no more issue. Good job.

**Comment by BrandonPacewic** [comment]
@LucaHaverty All done making changes?

**Pull request merged**
